### PR TITLE
Mathviews

### PR DESCRIPTION
**Pull request by Huy-Nguyen-HCI**
Add a front-end chat view with math formula and Wolframalpha suport.

### CODE REVIEW

File: app/views/chat.scala.html [hunk 1]
```diff
@@ -4,17 +4,22 @@
...
-        <p>Enter math formula</p>
-        <div class="col-xs-8 offset-xs-2" id="mathquill"></div>
+    <div id="wrap" class="row navbar navbar-fixed-bottom">
+        <div class="col-xs-10">
+            <div id="mathquill" class="textbox"></div>
+        </div>
     </div>
     <form id="answer-form" method="POST">
     </form>
+
```
Comment (fatlotus) [inline app/views/chat.scala.html]: format?

File: app/views/chat.scala.html [hunk 2]
```diff
@@ -4,17 +4,22 @@
...
 </div>
 
 @keyboard()
 
 <script type="text/javascript" src='@routes.Assets.versioned("javascripts/math_input.js")'></script>
+<script type="text/javascript">
```
Comment (fatlotus) [inline app/views/chat.scala.html]: ?

File: app/views/chat.scala.html [hunk 3]
```diff
@@ -1,20 +1,265 @@
...
 *@
 
-@main("Chat") {
-<div class="container">
-    <div id="wrap" class="row">
-        <p>Enter math formula</p>
-        <div class="col-xs-8 offset-xs-2" id="mathquill"></div>
+@main("Chat room") {
+<div class="main_section">
+    <div class="container" style="width: 100%; padding: 0;">
```
Comment (fatlotus) [inline app/views/chat.scala.html]: remove inline style?

File: app/views/chat.scala.html [hunk 4]
```diff
@@ -1,20 +1,265 @@
...
+        <div class="chat_container">
+            <div class="col-sm-3 chat_sidebar">
```
Comment (fatlotus) [inline app/views/chat.scala.html]: dashes instead of underscores

File: app/views/chat.scala.html [hunk 5]
```diff
@@ -1,20 +1,265 @@
...
+                <div class="row">
+                    <div id="custom-search-input">
+                        <div class="input-group col-md-12">
+                            <input type="text" class="  search-query form-control" placeholder="Conversation"/>
```
Comment (fatlotus) [inline app/views/chat.scala.html]: whitespace

File: app/views/chat.scala.html [hunk 6]
```diff
@@ -1,20 +1,265 @@
...
+                            <button class="btn btn-danger" type="button">
+                                <span class=" glyphicon glyphicon-search"></span>
```
Comment (fatlotus) [inline app/views/chat.scala.html]: whitespace

File: app/views/chat.scala.html [hunk 7]
```diff
@@ -1,20 +1,265 @@
...
+                            <li><a href="#">Another action</a></li>
+                            <li><a href="#">Something else here</a></li>
+                            <li><a href="#">Separated link</a></li>
+                        </ul>
+                    </div>
+                    <div class="member_list">
+                        <ul class="list-unstyled">
+                            <li class="left clearfix">
+                     <span class="chat-img pull-left">
+                     <img src="https://avatars0.githubusercontent.com/u/9096489?v=3&s=460" alt="User Avatar"
```
Comment (fatlotus) [inline app/views/chat.scala.html]: nit: &amp;

File: app/views/chat.scala.html [hunk 8]
```diff
@@ -1,20 +1,265 @@
...
+                            </li>
+                            <li><a href="#">Another action</a></li>
+                            <li><a href="#">Something else here</a></li>
+                            <li><a href="#">Separated link</a></li>
+                        </ul>
+                    </div>
+                    <div class="member_list">
+                        <ul class="list-unstyled">
+                            <li class="left clearfix">
+                     <span class="chat-img pull-left">
```
Comment (fatlotus) [inline app/views/chat.scala.html]: reindent

File: public/javascripts/math_input.js [hunk 9]
```diff
@@ -3,7 +3,24 @@
...
+		setupMathInput();
+	}
+	else {
+		$textarea = $("<textarea class='form-control' placeholder='type a message'></textarea>").appendTo('#inputArea');
+		closeKeyboard();
+	}
+  });
+});
+
+function setupMathInput() {
```
Comment (fatlotus) [inline public/javascripts/math_input.js]: var setupMathInput = function() {

File: public/javascripts/math_input.js [hunk 10]
```diff
@@ -15,64 +32,69 @@ $(document).ready(function() {
...
-    } else if($(this).data('action') == 'keystroke') {
-      mathField.keystroke($(this).data('content'));
-    } else if($(this).data('action') == 'switch-keys') {
-      console.log("'#keyboard .keypad:not(.' + $(this).data('content') + ')'");
-      $('#keyboard .keypad:not(.' + $(this).data('content') + ')').hide();
-      $('#keyboard .keypad.' + $(this).data('content')).show();
-    } else if($(this).data('action') == 'keyboard-hide'){
-      closeKeyboard();
-    }
+    if ( mathField ) {
```
Comment (fatlotus) [inline public/javascripts/math_input.js]: if (foobar) {

also below

File: public/javascripts/math_input.js [hunk 11]
```diff
@@ -15,64 +32,69 @@ $(document).ready(function() {
...
   $('#keyboard-mask').height($('#keyboard-wrapper').height());
+}
+
+
+function closeKeyboard() {
+  $('#keyboard-mask').slideUp();
+  $('#keyboard-wrapper').slideUp();
+}
+
+function openKeyboard() {
```
Comment (fatlotus) [inline public/javascripts/math_input.js]: reformat

File: app/views/chat.scala.html [hunk 12]
```diff
@@ -1,20 +1,265 @@
 @*
 * The chat interface
+* Source: http://bootsnipp.com/user/snippets/GQP0E
```
Comment (fatlotus) [inline app/views/chat.scala.html]: can't sign in

**Review by fatlotus — CHANGES_REQUESTED** [review]
The jQuery seems a little tricky -- is there a way to do it without dynamic CSS selectors?

**Comment by fatlotus** [comment]
This would be a lot of work, but it might be interesting to add Selenium tests or something for the client-side logic.

File: app/views/chat.scala.html (hunk 12 above)
Comment (Huy-Nguyen-HCI) [reply app/views/chat.scala.html]: This is just a plain view; we're not linking it with any back-end yet.

File: public/javascripts/math_input.js [hunk 13]
```diff
@@ -3,7 +3,24 @@
...
-$(document).ready(function() {
+$(document).ready( function() {
+  setupMathInput();
+  // switch between normal text and math input
+  $('#typeMath').change( function() {
+	$('#inputArea').html("");
+	if (this.checked) {
+		$textarea = $("<div id='mathquill'></div>").appendTo('#inputArea');
+		setupMathInput();
+	}
```
Comment (fatlotus) [inline public/javascripts/math_input.js]: } else {

File: app/views/chat.scala.html (hunk 12 above)
Comment (fatlotus) [reply app/views/chat.scala.html]: oh, no -- i meant the snippet! Maybe copy to gist or something?

File: public/javascripts/math-input.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+var setupMathInput = function() {
+  // math input configuration
+  var isMobile = window.matchMedia("only screen and (max-width: 760px)");
+  var MQElement = document.getElementById("mathquill");
+
+  var MQ = MathQuill.getInterface(2); // for backcompat
+
+  var mathField = MQ.MathField(MQElement, {});
+
+//  $('#keyboard .keypad:not(.default)').hide();
```
Comment (fatlotus) [inline public/javascripts/math-input.js]: If possible, strip out commented code from the CL -- we already have full history in Git.

File: public/javascripts/math-input.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+  $('#keyboard .key').click( function(event) {
+    event.preventDefault();
+
+    if (mathField)  {
+      if ($(this).data('action') == 'write') {
+            mathField.write($(this).data('content'));
+      }
```
Comment (fatlotus) [inline public/javascripts/math-input.js]: } else {

File: public/javascripts/math-input.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        $(this).parents('#keyboard').switchClass("trig", $(this).data("content"));
+        $(this).parents('#keyboard').switchClass("std", $(this).data("content"));
+      }
+      else if($(this).data('action') == 'keyboard-hide'){
+        closeKeyboard();
+      }
+
+      if (typeof $(this).data('stepback') !== 'undefined') {
+        for (var i = 0; i < parseInt($(this).data('stepback')); i++) {
+        mathField.keystroke('Left');
```
Comment (fatlotus) [inline public/javascripts/math-input.js]: indent?

File: public/javascripts/math-input.js [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+
+var closeKeyboard = function() {
+  $('#keyboard-mask').slideUp();
+  $('#keyboard-wrapper').slideUp();
+}
+
+var openKeyboard = function() {
+	$('#keyboard-wrapper').slideDown();
```
Comment (fatlotus) [inline public/javascripts/math-input.js]: Please use either tabs or spaces (throughout the file)

(Might be possible to auto-format in your editor)

**Review by fatlotus — CHANGES_REQUESTED** [review]
Almost there! :) LGTM, modulo some nits.

**Review by fatlotus — APPROVED** [review]

**Pull request merged**